### PR TITLE
Fixes #30833 - Add search by role to hosts

### PR DESCRIPTION
**Pull request by adiabramovitch**

### CODE REVIEW

File: app/models/concerns/foreman_ansible/host_managed_extensions.rb [hunk 1]
```diff
@@ -64,9 +64,20 @@ def import_host(*args)
...
+
+      def search_by_role(_key, operator, value)
+        conditions = sanitize_sql_for_conditions(["ansible_roles.name #{operator} ?", value_to_sql(operator, value)])
+        host_ids = ::Host::Managed.joins(:ansible_roles).where(conditions).distinct.pluck(:id)
+        hostgroup_ids = ::Hostgroup.unscoped.with_taxonomy_scope.joins(:ansible_roles).where(conditions).map(&:subtree_ids).flatten
+
+        conds = []
+        conds << "hosts.id IN(#{host_ids.join(',')})" if host_ids.present?
+        conds << "hosts.hostgroup_id IN(#{hostgroup_ids.uniq.join(',')})" if hostgroup_ids.present?
+        { conditions: conds.join(' OR ').presence || 'hosts.id < 0' }
```
Comment (ezr-ondrej) [inline app/models/concerns/foreman_ansible/host_managed_extensions.rb:76]:
```suggestion
        { conditions: conds.join(' OR ').presence || '1 = 0' }
```

**Review by ezr-ondrej — COMMENTED** [review]
Looks good, just a nit, 1=0 is prefered way of expressing `nothing` in sql :)

**Comment by ezr-ondrej** [comment]
Also, if we could have a test, that would be great, I'll work on the rubocop fix :)

File: app/models/concerns/foreman_ansible/host_managed_extensions.rb [hunk 2]
```diff
@@ -64,9 +64,20 @@ def import_host(*args)
...
+
+        conds = []
+        conds << "hosts.id IN(#{host_ids.join(',')})" if host_ids.present?
+        conds << "hosts.hostgroup_id IN(#{hostgroup_ids.uniq.join(',')})" if hostgroup_ids.present?
+        { conditions: conds.join(' OR ').presence || '1 = 0' }
+      end
     end
   end
-end
+  end
```
Comment (ezr-ondrej) [inline app/models/concerns/foreman_ansible/host_managed_extensions.rb:80]: This is probably not right :thinking:

File: test/unit/concerns/host_managed_extensions_test.rb [hunk 3]
```diff
@@ -62,4 +62,12 @@ class HostManagedExtensionsTest < ActiveSupport::TestCase
     FactoryBot.create(:host_ansible_role, :ansible_role_id => @role1.id, :position => 0, :host_id => host.id)
     host.all_ansible_roles.must_equal [@role2, @role1]
   end
+
+  test 'should find hosts with role2' do
+    host1 = FactoryBot.create(:host, :host_ansible_role => [@role2])
```
Comment (ezr-ondrej) [inline test/unit/concerns/host_managed_extensions_test.rb:67]: This should be
```suggestion
    host1 = FactoryBot.create(:host, :ansible_roles => [@role2])
```

Comment (adiabramovitch) [reply test/unit/concerns/host_managed_extensions_test.rb]: sorry, you are right :)

**Review by ezr-ondrej — APPROVED** [review]
Failures unrelated. Thanks @adiabramovitch ! :+1:

**Pull request merged**